File: python/src/indian_fakedata/utils/sft.py
"""
SFT Pair Builder (v2.1.0, item 7)

Mirrors the TypeScript implementation (src/utils/sft.ts): turns a
profile (and optionally its narrative documents) into instruction /
response pairs for supervised fine-tuning, in Alpaca style:
{"instruction", "input", "output", "source"}. Every response is
templated strictly from profile fields, so pairs are grounded by
construction — nothing is hallucinated.

Pure functions, no RNG: output is fully determined by the inputs.
"""


def _inr(n):
    # Indian digit grouping to match toLocaleString('en-IN'): 1,02,900
    digits = str(round(n))
    if len(digits) <= 3:
        return "\u20b9" + digits
    head, tail = digits[:-3], digits[-3:]
    groups = []
    while len(head) > 2:
        groups.insert(0, head[-2:])
        head = head[:-2]
    if head:
        groups.insert(0, head)
    return "\u20b9" + ",".join(groups) + "," + tail


def _persona_pairs(profile):
    name = "{} {}".format(profile.get("firstName", ""), profile.get("lastName", ""))
    first = profile.get("firstName", "")
    marital = profile.get("maritalStatus", "never_married")
    if marital == "married":
        n_kids = profile.get("numberOfChildren", 0)
        family = ("I am married to {} and have {} {}. ".format(
            profile.get("spouseName") or "my spouse", n_kids,
            "child" if n_kids == 1 else "children") +
            "We live in {} {}.".format(profile.get("areaType", ""),
                                       profile.get("district", "")))
    else:
        family = ("I am currently {} and live in {} {} with {} family members.".format(
            str(marital).replace("_", " "), profile.get("areaType", ""),
            profile.get("district", ""), profile.get("householdSize", 0)))
    second = profile.get("secondLanguage")
    cultural = profile.get("culturalProfile", {})
    return [
        {
            "instruction": "You are {}. Introduce yourself in the first person.".format(name),
            "input": "",
            "output": ("My name is {}. I am a {}-year-old {} from {}, {}. "
                       "I work as a {} and my mother tongue is {}.".format(
                           name, profile.get("age", 0), profile.get("gender", ""),
                           profile.get("district", ""), profile.get("state", ""),
                           profile.get("employmentSector", ""),
                           profile.get("motherTongue", ""))),
            "source": "persona",
        },
        {
            "instruction": "As {}, describe what you do for a living.".format(first),
            "input": "",
            "output": ("I work as a {} and earn about {} a month to support "
                       "my household of {}.".format(
                           profile.get("employmentSector", ""),
                           _inr(profile.get("annualIncomeINR", 0) / 12),
                           profile.get("householdSize", 0))),
            "source": "persona",
        },
        {
            "instruction": "As {}, describe your family.".format(first),
            "input": "",
            "output": family,
            "source": "persona",
        },
        {
            "instruction": "As {}, talk about your education and languages.".format(first),
            "input": "",
            "output": ("My highest education is {}. I speak {}{}.".format(
                str(profile.get("education", "")).replace("_", " "),
                profile.get("motherTongue", ""),
                " and {}".format(second) if second else "")),
            "source": "persona",
        },
        {
            "instruction": "As {}, how do you handle household money?".format(first),
            "input": "",
            "output": ("Our household spends about {} a month. I {}.".format(
                _inr(profile.get("monthlyExpenditureINR", 0)),
                "save money diligently" if cultural.get("savingsOrientation", 0) > 60
                else "spend within our means")),
            "source": "persona",
        },
    ]


def _narrative_qa(doc_type):
    makers = {
        "loan_application": (
            "Who is this loan application from, and which bank is it addressed to?",
            lambda p: ("It is a personal loan application from {} {} of {}, "
                       "addressed to {}.".format(
                           p.get("firstName", ""), p.get("lastName", ""),
                           p.get("district", ""), p.get("bankName", "")))),
        "medical_consultation": (
            "Who is the patient in this consultation record and where are they from?",
            lambda p: ("The patient is {} {}, a {}-year-old from {}, {}.".format(
                p.get("firstName", ""), p.get("lastName", ""), p.get("age", 0),
                p.get("district", ""), p.get("state", "")))),
        "hinglish_conversation": (
            "Who are the speakers in this chat and what language mix do they use?",
            lambda p: ("{} chats with a friend in Hinglish, mixing {} and English.".format(
                p.get("firstName", ""), p.get("motherTongue", "")))),
        "ration_card_application": (
            "Whose ration card application is this and what type is requested?",
            lambda p: ("It belongs to {} {} of {}, holding a {} card.".format(
                p.get("firstName", ""), p.get("lastName", ""),
                p.get("district", ""), p.get("rationCardType", "")))),
        "school_enrollment": (
            "Who is being enrolled and in which district?",
            lambda p: ("{} {} is being enrolled for school in {} district.".format(
                p.get("firstName", ""), p.get("lastName", ""),
                p.get("district", "")))),
        "resume": (
            "Whose resume is this and what work do they do?",
            lambda p: ("It is the resume of {} {}, working as a {}.".format(
                p.get("firstName", ""), p.get("lastName", ""),
                p.get("employmentSector", "")))),
        "customer_support_chat": (
            "Who is the customer in this support chat and which bank helps them?",
            lambda p: ("The customer is {} {}, helped by {} customer care.".format(
                p.get("firstName", ""), p.get("lastName", ""),
                p.get("bankName", "")))),
    }
    return makers.get(doc_type)


def build_sft_pairs(profile, narratives=None):
    """
    Build grounded SFT pairs for one profile, plus one
    reading-comprehension pair per narrative document when provided.
    """
    pairs = _persona_pairs(profile)
    for doc in narratives or []:
        maker = _narrative_qa(doc.get("type"))
        if not maker:
            continue
        question, answer_fn = maker
        pairs.append({"instruction": question,
                      "input": doc.get("content", ""),
                      "output": answer_fn(profile),
                      "source": doc.get("type")})
    return pairs


def sft_pairs_to_jsonl(pairs):
    """Serialize pairs as JSONL, one example per line."""
    import json
    if not pairs:
        return ""
    return "\n".join(json.dumps(p, ensure_ascii=False) for p in pairs) + "\n"

File: src/utils/sft.ts
/**
 * SFT Pair Builder (v2.1.0, item 7)
 *
 * Turns a profile (and optionally its narrative documents) into
 * instruction/response pairs for supervised fine-tuning, in Alpaca style:
 * { instruction, input, output, source }. Every response is templated
 * strictly from profile fields, so pairs are grounded by construction —
 * nothing is hallucinated.
 *
 * Pure functions, no RNG: output is fully determined by the inputs.
 */

import type { DemographicProfile } from '../types.js';
import type { NarrativeDocument } from './narrative.js';

/** One supervised fine-tuning example */
export interface SFTPair {
  instruction: string;
  input: string;
  output: string;
  /** Where this pair came from: 'persona' or a narrative doc type */
  source: string;
}

function inr(n: number): string {
  return `₹${Math.round(n).toLocaleString('en-IN')}`;
}

function personaPairs(profile: DemographicProfile): SFTPair[] {
  const name = `${profile.firstName} ${profile.lastName}`;
  return [
    {
      instruction: `You are ${name}. Introduce yourself in the first person.`,
      input: '',
      output:
        `My name is ${name}. I am a ${profile.age}-year-old ${profile.gender} ` +
        `from ${profile.district}, ${profile.state}. I work as a ${profile.employmentSector} ` +
        `and my mother tongue is ${profile.motherTongue}.`,
      source: 'persona',
    },
    {
      instruction: `As ${profile.firstName}, describe what you do for a living.`,
      input: '',
      output:
        `I work as a ${profile.employmentSector} and earn about ` +
        `${inr(profile.annualIncomeINR / 12)} a month to support ` +
        `my household of ${profile.householdSize}.`,
      source: 'persona',
    },
    {
      instruction: `As ${profile.firstName}, describe your family.`,
      input: '',
      output:
        profile.maritalStatus === 'married'
          ? `I am married to ${profile.spouseName ?? 'my spouse'} and have ` +
            `${profile.numberOfChildren} ${profile.numberOfChildren === 1 ? 'child' : 'children'}. ` +
            `We live in ${profile.areaType} ${profile.district}.`
          : `I am currently ${profile.maritalStatus.replace(/_/g, ' ')} and live ` +
            `in ${profile.areaType} ${profile.district} with ${profile.householdSize} family members.`,
      source: 'persona',
    },
    {
      instruction: `As ${profile.firstName}, talk about your education and languages.`,
      input: '',
      output:
        `My highest education is ${profile.education.replace(/_/g, ' ')}. ` +
        `I speak ${profile.motherTongue}` +
        (profile.secondLanguage ? ` and ${profile.secondLanguage}` : '') +
        `.`,
      source: 'persona',
    },
    {
      instruction: `As ${profile.firstName}, how do you handle household money?`,
      input: '',
      output:
        `Our household spends about ${inr(profile.monthlyExpenditureINR)} a month. ` +
        `I ${profile.culturalProfile.savingsOrientation > 60 ? 'save money diligently' : 'spend within our means'}.`,
      source: 'persona',
    },
  ];
}

const NARRATIVE_QUESTIONS: Record<string, (p: DemographicProfile) => { q: string; a: string }> = {
  loan_application: p => ({
    q: 'Who is this loan application from, and which bank is it addressed to?',
    a: `It is a personal loan application from ${p.firstName} ${p.lastName} of ${p.district}, addressed to ${p.bankName}.`,
  }),
  medical_consultation: p => ({
    q: 'Who is the patient in this consultation record and where are they from?',
    a: `The patient is ${p.firstName} ${p.lastName}, a ${p.age}-year-old from ${p.district}, ${p.state}.`,
  }),
  hinglish_conversation: p => ({
    q: 'Who are the speakers in this chat and what language mix do they use?',
    a: `${p.firstName} chats with a friend in Hinglish, mixing ${p.motherTongue} and English.`,
  }),
  ration_card_application: p => ({
    q: 'Whose ration card application is this and what type is requested?',
    a: `It belongs to ${p.firstName} ${p.lastName} of ${p.district}, holding a ${p.rationCardType} card.`,
  }),
  school_enrollment: p => ({
    q: 'Who is being enrolled and in which district?',
    a: `${p.firstName} ${p.lastName} is being enrolled for school in ${p.district} district.`,
  }),
  resume: p => ({
    q: 'Whose resume is this and what work do they do?',
    a: `It is the resume of ${p.firstName} ${p.lastName}, working as a ${p.employmentSector}.`,
  }),
  customer_support_chat: p => ({
    q: 'Who is the customer in this support chat and which bank helps them?',
    a: `The customer is ${p.firstName} ${p.lastName}, helped by ${p.bankName} customer care.`,
  }),
};

/**
 * Build grounded SFT pairs for one profile, plus one reading-comprehension
 * pair per narrative document when provided.
 */
export function buildSFTPairs(
  profile: DemographicProfile,
  narratives: NarrativeDocument[] = []
): SFTPair[] {
  const pairs = personaPairs(profile);
  for (const doc of narratives) {
    const maker = NARRATIVE_QUESTIONS[doc.type];
    if (!maker) continue;
    const { q, a } = maker(profile);
    pairs.push({ instruction: q, input: doc.content, output: a, source: doc.type });
  }
  return pairs;
}

/** Serialize pairs as JSONL, one example per line. */
export function sftPairsToJsonl(pairs: SFTPair[]): string {
  return pairs.map(p => JSON.stringify(p)).join('\n') + (pairs.length > 0 ? '\n' : '');
}
